docs/optimization/feature_selection.md

```diff
@@ -0,0 +1,421 @@
+  name: python3
+---
+
+```{code-cell} ipython3
+:tags: []
+
+%load_ext autoreload
+%autoreload 2
+```
+
+# Feature Selection
+
+In a dataset, not every feature might be useful. Some could be noise, and some could show a strong relationship with the target variable. Often, using all the features decreases a model's performance and increases training time. Therefore, it's best to select the best subset of features that are relevant to the target variable before we train a model. This process is called feature selection. 
+
+With supervised learning, feature selection has 3 main categories.
+
+#### 1. Filter method
+#### 2. Wrapper method
+#### 3. Embedded method
+
+In this tutorial, we will go over what those 3 categories are, what methods are under the 3 categories, and how to implement those with sklearn. So let's first bring in toy datasets from the sklearn package.
+
+```{code-cell} ipython3
+:tags: []
+
+import pandas as pd
+import numpy as np
+from sklearn.model_selection import train_test_split
+from sklearn.datasets import load_diabetes, load_breast_cancer
+
+# Preparing regression dataset
+diabetes = load_diabetes(as_frame=True)
+X_reg, y_reg = diabetes.data, diabetes.target
+X_reg_train, X_reg_test, y_reg_train, y_reg_test = train_test_split(X_reg, y_reg, 
+                                                                    test_size=0.2, 
+                                                                    random_state=0)
+
+# Preparing classification dataset
+breast_cancer = load_breast_cancer(as_frame=True)
+X_clf, y_clf = breast_cancer.data, breast_cancer.target
+X_clf_train, X_clf_test, y_clf_train, y_clf_test = train_test_split(X_clf, y_clf, 
+                                                                    test_size=0.2, 
+                                                                    random_state=0)
+```
+
+```{code-cell} ipython3
+:tags: []
+
+# Example rows of regression dataset
+X_reg_train.head(4)
+```
+
+```{code-cell} ipython3
+:tags: []
+
+# Example rows of classification dataset
+X_clf_train.head(4)
+```
+
++++ {"tags": []}
+
+## 1. Filter methods
+
+Filter methods involve testing and selecting features based on univariate statistical tests on predictor variables or relationships between predictor and target variables. Thus, they tend to be fast and less computationally expensive than the other two methods. However, they usually fail to find the best subset of features and give small model improvements.
+
+Several techniques are
+1. Quasi Constant
+2. Pearson's Correlation Coefficient
+3. Spearman's Rank Correlation
+4. ANOVA F-test
+5. Chi-Square Test
+6. Mutual Information
+
++++ {"tags": []}
+
+### 1.1. Quasi Constant
+
+A Quasi constant feature is a feature whose majority are the same values. If all or most of the values are the same, it won't be useful for prediction. So after you set a minimum threshold for variance, you remove features based on whether their variances exceed the threshold or not. 
```

Comment:
reframe this sentence -> majority of the samples have same value of this feature